### PR TITLE
server: adds minimal Secret Service server implementation

### PR DESCRIPTION
**Pull request by warusadura**
This change adds the *minimal server side interface implementation of the freedesktop Secret Service API. The implementation is NOT complete. This change provides a foundation for adding further, more accurate implementations on top of it.

*minimal: Included all the Secret Service API method signatures without any implementation (todo!). Also, added a simple main().

Secret Service API spec: https://specifications.freedesktop.org/secret-service-spec/latest/

### CODE REVIEW

**Comment by warusadura** [comment]
I think this pr is still too large :) I'll try to make this smaller.

**Comment by bilelmoussaoui** [comment]
I would expect a first PR that just adds service.rs with the bare minimum required methods. Or all the files but without any implementations and just todo! calls

**Comment by warusadura** [comment]
before: `733` lines, now: `324`. I have removed all the implementations. I think this provides a good base now :)

File: server/src/daemon/error.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Errors
+
+#[derive(zbus::DBusError, Debug)]
+#[zbus(prefix = "org.freedesktop.Secret.Error")]
+pub enum ServiceError {
```
Comment (bilelmoussaoui) [inline server/src/daemon/error.rs:5]: The exact same error is defined https://github.com/bilelmoussaoui/oo7/blob/main/client/src/dbus/error.rs#L6 already, make use of that instead

File: server/src/daemon/error.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    NoSession,
+    // Collection/Item does not exist.
+    NoSuchObject,
+}
+
+impl From<oo7::portal::Error> for ServiceError {
+    fn from(value: oo7::portal::Error) -> Self {
+        Self::ZBus(zbus::Error::Failure(value.to_string()))
+    }
+}
```
Comment (bilelmoussaoui) [inline server/src/daemon/error.rs:21]: This is not used now

File: server/src/daemon/item.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// org.freedesktop.Secret.Item
+
+use oo7::dbus::api::SecretInner;
+use zbus::zvariant::{ObjectPath, OwnedObjectPath};
+
+use super::Result;
+
+#[derive(Debug)]
+pub struct Item {
+    _path: OwnedObjectPath,
```
Comment (bilelmoussaoui) [inline server/src/daemon/item.rs:10]: Drop the unused fields

File: server/src/daemon/item.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+#[zbus::interface(name = "org.freedesktop.Secret.Item")]
+impl Item {
+    pub async fn delete(&self) -> Result<ObjectPath> {
+        todo!()
+    }
+
+    #[zbus(name = "GetSecret")]
```
Comment (bilelmoussaoui) [inline server/src/daemon/item.rs:19]: The methods should name the in & out arguments so they are correctly displayed on apps like d-spy

File: server/src/daemon/service.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    pub async fn set_alias(&self, _name: &str, _collection: ObjectPath<'_>) -> Result<()> {
+        todo!()
+    }
+}
+
+impl Service {
+    pub async fn new() -> Self {
+        Self {
+            cnx: Default::default(),
```
Comment (bilelmoussaoui) [inline server/src/daemon/service.rs:84]: Why the cnx should be a mutex ? Just remove the new method and have run create the connection itself

File: server/src/daemon/service.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+
+    pub async fn run(self) -> Result<()> {
+        let cnx = zbus::connection::Builder::session()?
+            .name("org.freedesktop.secrets.Devel")?
+            .build()
+            .await?;
+        *self.cnx.lock().unwrap() = Some(cnx.clone());
+        cnx.object_server()
```
Comment (A6GibKm) [inline server/src/daemon/service.rs:94]: Can this be done before the previous line and then not have to do the `clone()`?

File: server/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "oo7-daemon"
```
Comment (bilelmoussaoui) [inline server/Cargo.toml:2]: This needs to be added to the CI as well so that it always passes before merging any further things

File: Cargo.lock [hunk 8]
```diff
@@ -813,6 +813,16 @@ version = "0.4.14"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "78b3ae25bc7c8c38cec158d1f2757ee79e9b3740fbc7ccf0e59e4b08d793fa89"
 
+[[package]]
```
Comment (bilelmoussaoui) [inline Cargo.lock:816]: The updates to Cargo.lock are irrelevant, please split into a separate commit.

**Review by bilelmoussaoui — COMMENTED** [review]
much better!

File: server/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+mod daemon;
```
Comment (bilelmoussaoui) [inline server/src/main.rs:1]: I don't understand the reason for having this in a separate mod; it is not like we are going to expose it as a separate library anyways. So drop the folder. initially, i was suggesting to name "server" as "daemon" instead of adding an extra mod.

File: server/src/daemon/service.rs (hunk 5 above)
Comment (A6GibKm) [reply server/src/daemon/service.rs:84]: Anotherway of saying this,  the `cnx` should only be set once. There is no need to set a `default` dummy value.

File: server/src/daemon/error.rs (hunk 1 above)
Comment (warusadura) [reply server/src/daemon/error.rs]: done, thanks for the prompt review.

File: server/src/daemon/item.rs (hunk 3 above)
Comment (warusadura) [reply server/src/daemon/item.rs]: done.

File: server/src/daemon/error.rs (hunk 2 above)
Comment (warusadura) [reply server/src/daemon/error.rs]: removed.

File: server/src/daemon/service.rs (hunk 5 above)
Comment (warusadura) [reply server/src/daemon/service.rs]: understood, thanks! @A6GibKm

File: server/src/main.rs (hunk 9 above)
Comment (warusadura) [reply server/src/main.rs]: sorry for the mix up. this is fixed now.

File: server/src/daemon/item.rs (hunk 4 above)
Comment (warusadura) [reply server/src/daemon/item.rs]: I couldn't find a way to name `IN` arguments. but, I have named all the `OUT` args.

File: server/Cargo.toml (hunk 7 above)
Comment (warusadura) [reply server/Cargo.toml]: looking into this now.

Comment (warusadura) [reply server/Cargo.toml]: done.

File: server/src/daemon/item.rs (hunk 4 above)
Comment (bilelmoussaoui) [reply server/src/daemon/item.rs]: The in argument are the name you give to the function parameters

File: Cargo.lock (hunk 8 above)
Comment (bilelmoussaoui) [reply Cargo.lock:816]: Please update the commit to just say "update dependencies"

**Comment by bilelmoussaoui** [comment]
Please update the commits like the following:
- Add a minimal Secret Service server implementation
and include the CI bits in it
- Rename the update dependencies one to "Update dependencies" and put it before the server impl one
otherwise good to be merged!

File: server/src/daemon/item.rs (hunk 4 above)
Comment (warusadura) [reply server/src/daemon/item.rs]: got it, thanks!

File: Cargo.lock (hunk 8 above)
Comment (warusadura) [reply Cargo.lock:816]: done, updated commit names and order.

**Comment by A6GibKm** [comment]
This is not necessarily something that should be addressed in this PR, but I am not particularly happy with putting the project in a `daemon` folder, the reason being that there are actually two daemons, the portal and the fdo secrets interface provider.

**Comment by warusadura** [comment]
> This is not necessarily something that should be addressed in this PR, but I am not particularly happy with putting the project in a `daemon` folder, the reason being that there are actually two daemons, the portal and the fdo secrets interface provider.

I'm happy to rename the project dir. wdyt @bilelmoussaoui 

> fdo secrets interface provider.

this is not done yet right (as in not in oo7)? is there a place I can read more about this? I don't know anything about fdo :) @A6GibKm

**Comment by A6GibKm** [comment]
https://www.freedesktop.org/wiki/Specifications/secret-storage-spec/secrets-api-0.1.html freedesktop being a reference to it being a freedesktop spec. 

Well `server` has the same problem. Again, I don't think this should be addressed here (or if it does need addressing), but it is something to consider for the long run.

**Comment by warusadura** [comment]
my bad I was thinking about FIDO :D

**Comment by bilelmoussaoui** [comment]
> Well server has the same problem. Again, I don't think this should be addressed here (or if it does need addressing), but it is something to consider for the long run.

Tbh, the portal is already clearly named. Server matches the client side, so i would go with that. I don't think it is really important at the end, no one will ever deal with it

File: daemon/src/service.rs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    pub async fn run(self) -> Result<()> {
+        let connection = zbus::connection::Builder::session()?
+            .name("org.freedesktop.secrets.Devel")?
+            .build()
+            .await?;
+        connection
+            .object_server()
+            .at(oo7::dbus::api::Service::PATH.unwrap(), self)
+            .await?;
+        Ok(())
```
Comment (bilelmoussaoui) [inline daemon/src/service.rs:92]:
```suggestion
    pub async fn run() -> Result<()> {
        let connection = zbus::connection::Builder::session()?
            .name("org.freedesktop.secrets.Devel")?
            .build()
            .await?;
        connection
            .object_server()
            .at(oo7::dbus::api::Service::PATH.unwrap(), Self::default())
            .await?;
        Ok(())
```

and then just call Service::run().await?; instead.

File: daemon/src/service.rs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    pub async fn set_alias(&self, _name: &str, _collection: ObjectPath<'_>) -> Result<()> {
+        todo!()
+    }
+}
+
+impl Service {
+    pub async fn run(self) -> Result<()> {
+        let connection = zbus::connection::Builder::session()?
+            .name("org.freedesktop.secrets.Devel")?
```
Comment (bilelmoussaoui) [inline daemon/src/service.rs:85]: why is this name hardcoded? it is accessible from oo7::dbus::api like the path. If it is for making debugging easier, we would need a proper infrastructure for it, which can be added in a separate commit.

File: daemon/src/service.rs (hunk 10 above)
Comment (warusadura) [reply daemon/src/service.rs]: done, thanks!

File: daemon/src/service.rs (hunk 11 above)
Comment (warusadura) [reply daemon/src/service.rs]: yeah, done.

**Comment by warusadura** [comment]
included a README file. I'll add license information later. I'm suppose to follow [this](https://wiki.gnome.org/Foundation/OpenSourceGuidelines).

**Comment by bilelmoussaoui** [comment]
nice, thanks!
